checker/src/org/checkerframework/checker/index/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,153 @@
+        // Do reassignment things here.
+        super.updateForAssignment(n, val);
+
+        // This code determines the list of dependencies in types that are to be invalidated
+        SideEffectKind sideEffectKind;
+        if (n instanceof LocalVariableNode
+                && n.getType().getKind() == TypeKind.ARRAY
+                && !ElementUtils.isEffectivelyFinal(((LocalVariableNode) n).getElement())) {
+            sideEffectKind = UpperBoundUtil.SideEffectKind.LOCAL_VAR_REASSIGNMENT;
+        } else if (n instanceof FieldAccessNode
+                && !ElementUtils.isEffectivelyFinal(((FieldAccessNode) n).getElement())) {
+            if (n.getType().getKind() == TypeKind.ARRAY) {
+                sideEffectKind = UpperBoundUtil.SideEffectKind.ARRAY_FIELD_REASSIGNMENT;
+            } else if (!n.getType().getKind().isPrimitive()) {
+                sideEffectKind = UpperBoundUtil.SideEffectKind.NON_ARRAY_FIELD_REASSIGNMENT;
+            } else {
+                return; // No side effect to check
+            }
+        } else {
+            return; // No side effect to check
+        }
+
+        FlowExpressions.Receiver rec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), n);
+        clearFromStore(rec, n, sideEffectKind);
+    }
+
+    /**
+     * Clears receivers from the store whose current value may be affected by the side effect.
+     *
+     * @param reassignedVar reassigned variable or null
+     * @param n current node
+     * @param sideEffect kind of side effect
+     */
+    private void clearFromStore(Receiver reassignedVar, Node n, SideEffectKind sideEffect) {
+        TreePath path = analysis.getTypeFactory().getPath(n.getTree());
+        Set<Entry<? extends Receiver, CFValue>> receiverAnnotationEntry = new HashSet<>();
+        receiverAnnotationEntry.addAll(localVariableValues.entrySet());
+        receiverAnnotationEntry.addAll(methodValues.entrySet());
+        receiverAnnotationEntry.addAll(classValues.entrySet());
+        receiverAnnotationEntry.addAll(fieldValues.entrySet());
+        receiverAnnotationEntry.addAll(arrayValues.entrySet());
+
+        for (Entry<? extends Receiver, CFValue> entry : receiverAnnotationEntry) {
+            Receiver r = entry.getKey();
+            try {
+                for (Receiver dependent :
+                        UpperBoundUtil.getDependentReceivers(
+                                entry.getValue().getAnnotations(),
+                                path,
+                                (UpperBoundAnnotatedTypeFactory) analysis.getTypeFactory())) {
+                    if (UpperBoundUtil.isSideEffected(dependent, reassignedVar, sideEffect)
+                            != UpperBoundUtil.SideEffectError.NO_ERROR) {
+                        this.clearValue(r);
+                    }
+                }
+            } catch (NullPointerException e) {
+                // This happens in the second round of type-checking of a finally block,
```

Comment:
Is it `path` that is null?  If so, then this is related to #1520.  Also, if that's the case, check if `path` is  null and just return if it is.  (And move the comment there.)

(Technically, it's not the second round of type-checking, it's performing the dataflow analysis on the finally block that has been duplicated to represent the exceptional case.)

Reply:
`path` is null, but that's not the direct cause of the NPE. Checking if `path` is `null` is too broad - I tried that, and it introduces additional unsoundness (possibly related to #1520) as `path` is `null` sometimes even when no NPE is thrown.